ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -0,0 +1,391 @@
+      rb_sys_fail(
+        "Could not start CpuAndWallTimeWorker: Could not re-install pre-existing SIGPROF signal handler. " \
+        "This may break the component had installed it."
+      );
+    }
+
+    rb_raise(rb_eRuntimeError, "Could not start CpuAndWallTimeWorker: There's a pre-existing SIGPROF signal handler");
+  }
+}
+
+static void remove_sigprof_signal_handler(void) {
+  struct sigaction signal_handler_config = {
+    .sa_handler = SIG_DFL, // Reset back to default
+    .sa_flags = SA_RESTART // TODO: Unclear if this is actually needed/does anything at all
+  };
+  sigemptyset(&signal_handler_config.sa_mask);
+
+  if (sigaction(SIGPROF, &signal_handler_config, NULL) != 0) rb_sys_fail("Failure while removing the signal handler");
+}
+
+static void block_sigprof_signal_handler_from_running_in_current_thread(void) {
+  sigset_t signals_to_block;
+  sigemptyset(&signals_to_block);
+  sigaddset(&signals_to_block, SIGPROF);
+  pthread_sigmask(SIG_BLOCK, &signals_to_block, NULL);
+}
+
+static void handle_sampling_signal(DDTRACE_UNUSED int _signal, DDTRACE_UNUSED siginfo_t *_info, DDTRACE_UNUSED void *_ucontext) {
+  if (!ruby_thread_has_gvl_p()) {
+    return; // Not safe to enqueue a sample from this thread
+  }
+
+  // We implicitly assume there can be no concurrent nor nested calls to handle_sampling_signal because
+  // a) we get triggered using SIGPROF, and the docs state second SIGPROF will not interrupt an existing one
+  // b) we validate we are in the thread that has the global VM lock; if a different thread gets a signal, it will return early
+  //    because it will not have the global VM lock
+  // TODO: Validate that this does not impact Ractors
+
+  // Note: rb_postponed_job_register_one ensures that if there's a previous sample_from_postponed_job queued for execution
+  // then we will not queue a second one. It does this by doing a linear scan on the existing jobs; in the future we
+  // may want to implement that check ourselves.
+
+  // TODO: Do something with result (potentially update tracking counters?)
+  /*int result =*/ rb_postponed_job_register_one(0, sample_from_postponed_job, NULL);
+}
+
+// The actual sampling trigger loop always runs **without** the global vm lock.
+static void *run_sampling_trigger_loop(void *state_ptr) {
+  struct cpu_and_wall_time_worker_state *state = (struct cpu_and_wall_time_worker_state *) state_ptr;
+
+  struct timespec time_between_signals = {.tv_nsec = 10 * 1000 * 1000 /* 10ms */};
+
+  while (state->should_run) {
+    // TODO: This is still a placeholder for a more complex mechanism. In particular:
+    // * We want to signal a particular thread or threads, not the process in general
+    // * We want to track if a signal landed on the thread holding the global VM lock and do something about it
+    // * We want to do more than having a fixed sampling rate
+
+    kill(getpid(), SIGPROF);
+    nanosleep(&time_between_signals, NULL);
```

Comment:
Are we ok with a small drift, given `nanosleep` won't called exactly at the end of the last 10ms interval?

Reply:
Nice catch, and yes!

As the many TODOs indicate, this area is one where I plan to still poke a lot; my plan is to adopt the same dynamic sampling approach we use for the old profiler, where the rate of sampling is adjusted to keep the overhead down.

I may decide to keep the fixed rate sampling as an option for advanced customers... but I'd still need to make sure we're not overwhelming the Ruby VM (e.g. if we're sampling so often that the VM is not able to make reasonable progress).